spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/AbstractBinder.java

```diff
@@ -373,60 +353,25 @@ public void afterPropertiesSet() throws Exception {
 		return doBindConsumer(name, group, inputChannel, properties);
 	}
 
-	protected abstract Binding<MessageChannel> doBindConsumer(String name, String group, MessageChannel inputChannel, Properties properties);
-
-	/**
-	 * Create a producer for the named channel and bind it to the binder. Synchronized to avoid creating multiple
-	 * instances.
-	 * @param name The name.
-	 * @param channelName The name of the channel to be created, and registered as bean.
-	 * @param properties The properties.
-	 * @return The channel.
-	 */
-	protected synchronized MessageChannel doBindDynamicProducer(String name, String channelName,
-			Properties properties) {
-		MessageChannel channel = this.directChannelProvider.lookupSharedChannel(channelName);
-		if (channel == null) {
-			try {
-				channel = this.directChannelProvider.createAndRegisterChannel(channelName);
-				bindProducer(name, channel, properties);
-			}
-			catch (RuntimeException e) {
-				destroyCreatedChannel(channelName, channel);
-				throw new BinderException(
-						"Failed to bind dynamic channel '" + name + "' with properties " + properties, e);
-			}
-		}
-		return channel;
-	}
-
-	private void destroyCreatedChannel(String name, MessageChannel channel) {
-		BeanFactory beanFactory = this.applicationContext.getBeanFactory();
-		if (beanFactory.containsBean(name)) {
-			if (beanFactory instanceof DefaultListableBeanFactory) {
-				((DefaultListableBeanFactory) beanFactory).destroySingleton(name);
-			}
-		}
-	}
+	protected abstract Binding<T> doBindConsumer(String name, String group, T inputChannel, Properties properties);
 
-	@Override
-	public void unbind(Binding<MessageChannel> binding) {
-		binding.stop();
+	protected void notifyUnbind(DefaultBinding<T> binding) {
 		this.bindings.remove(binding);
 		afterUnbind(binding);
 	}
 
-	protected void afterUnbind(Binding<MessageChannel> binding) {
+	protected void afterUnbind(DefaultBinding<T> binding) {
 	}
 
-	protected void addBinding(Binding<MessageChannel> binding) {
+	protected void addBinding(DefaultBinding<T> binding) {
 		this.bindings.add(binding);
 	}
 
-	protected void stopBindings() {
-		for (Lifecycle bean : this.bindings) {
+	@Override
+	public void destroy() throws Exception {
+		for (Binding<?> binding : this.bindings) {
 			try {
-				bean.stop();
+				binding.unbind();
```

Comment:
I don't think that would be good. For one thing, the Input/OutputBindingLifecycle allows binding/unbinding whenever the application is started or stopped (which in principle can happen multiple times during its lifecycle). This is redundant and needs to be removed on the long run, however it currently ensures that dynamic bindings are unbound as well (we need to fix that).